### PR TITLE
Create design document for constant update Q4 2018.

### PR DESCRIPTION
**Pull request by lrhn**
@eernstg Can you check if I have missed any changes?

### CODE REVIEW

File: accepted/future-releases/constant-update-2018/feature-specification.md [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+The constructor can still be invoked as a non-constant, though.
+
+### Clarifications
+
+The new specification clarifies that type variables cannot be used 
+as compile-time constant types, 
+and the type arguments to const constructor invocations and constant literals
+must be compile-time constant types.
+
+## Implementation Requirements
```
Comment (mit-mit) [inline accepted/future-releases/constant-update-2018/feature-specification.md]: This is expected to go into a second document named `implementation-plan.md`, see:
https://github.com/dart-lang/language/tree/master/accepted/2.1/super-mixins

File: accepted/future-releases/constant-update-2018/feature-specification.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+We also want to clean that up and give a clean, statically checkable
+definition of potentially constant expressions.
+
+We do not want to change the behavior of existing constant
+or potentially constant expressions.
+
+## Design
+
+The precise defintion of the changes [has landed](https://dart-review.googlesource.com/36220)
+in the language specification.
```
Comment (eernstg) [inline accepted/future-releases/constant-update-2018/feature-specification.md]: The CL links to the commit and the commit links to the CL, so they're essentially providing the same information. However, the link https://github.com/dart-lang/sdk/commit/0808d29b4e33c303bf40524407a6e331c7a439f8 seems to be slightly more direct in terms of what the readers of this document would actually need.

File: accepted/future-releases/constant-update-2018/feature-specification.md [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+`x` is a `String`, so there is no possible valid constant invocation of this
+constructor.
+The constructor can still be invoked as a non-constant, though.
+
+### Clarifications
+
+The new specification clarifies that type variables cannot be used 
+as compile-time constant types, 
+and the type arguments to const constructor invocations and constant literals
+must be compile-time constant types.
```
Comment (eernstg) [inline accepted/future-releases/constant-update-2018/feature-specification.md]: Maybe mention that, as a new thing, a type variable from the enclosing class used as an expression (not as a type) is a constant expression? With the text in lines 127-130 in mind, that seems to be enough of a near-contradiction to make it natural to mention.

Comment (eernstg) [reply accepted/future-releases/constant-update-2018/feature-specification.md]: Checking the current text, this seems to have disappeared again?:
```
\item An identifier denoting a type parameter of a generic class is a potentially constant expression when it occurs in the initializer list of a const constructor of that class.
```

Comment (lrhn) [reply accepted/future-releases/constant-update-2018/feature-specification.md]: It did. And thinking about it more, I think that was a wise move.
It makes it easier to talk about constant expressions evaluating to `Type` objects when type variables in a constant constructor initializer list are not included.

File: accepted/future-releases/constant-update-2018/feature-specification.md (hunk 1 above)
Comment (lrhn) [reply accepted/future-releases/constant-update-2018/feature-specification.md]: Will do.

File: accepted/future-releases/constant-update-2018/feature-specification.md (hunk 2 above)
Comment (lrhn) [reply accepted/future-releases/constant-update-2018/feature-specification.md]: Will use that instead. Thanks.

File: accepted/future-releases/constant-update-2018/feature-specification.md [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+The other branch must also be a potenatially constant expression.
+
+### Potentially Constant Expressions
+
+Potentially constant expressions were previously defined in terms of 
+"expressions that would be constant if constructor arguments are replaced with 
+constants of types matching the context", but something was a constant
+expression if it evaluated to a value without throwing. 
+
+That was not a functional definition,
```
[the remark this answers is not in the document]
Comment (lrhn) [reply accepted/future-releases/constant-update-2018/feature-specification.md]: Reworded.

File: accepted/future-releases/constant-update-2018/implementation-plan.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
+# Implementation Plan for the Q4 2018 Constant Update
+
+Relevant documents:
+ - `Tracking issue` - TBD
```
Comment (mit-mit) [inline accepted/future-releases/constant-update-2018/implementation-plan.md]: I created https://github.com/dart-lang/language/issues/60

File: accepted/future-releases/constant-update-2018/implementation-plan.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+When all SDK tools have done so, 
+the flag is removed from the experiments flag definition file.
+
+## Timeline
+
+Completion goals for the phases:
+
+- Phase 0: Mid November 2018
+- Phase 1: Early December 2018
+- Phase 2: Mid December 2018
```
Comment (mit-mit) [inline accepted/future-releases/constant-update-2018/implementation-plan.md]: I don't see a phase 3 above?

File: accepted/future-releases/constant-update-2018/feature-specification.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+## Design
+
+The precise defintion of the changes [has landed](dart-lang/sdk@0808d29)
+in the language specification. 
+The following section summarizes the changes.
+
+### Cast Operator
+An expression of the form `e as T` is accepted as a potentially 
+and compile-time constant experssion
```
Comment (mit-mit) [inline accepted/future-releases/constant-update-2018/feature-specification.md]: s/experssion/expression/

File: accepted/future-releases/constant-update-2018/feature-specification.md [hunk 8]
```diff
@@ -0,0 +1,130 @@
+# Design Document for Constant Specification Update 2018.
+Author: lrn@google.com (@lrhn)
+
+## Motivation and Scope
```
Comment (mit-mit) [inline accepted/future-releases/constant-update-2018/feature-specification.md]: Can we give a concrete example of code that illustrates some of these issues?

Comment (lrhn) [reply accepted/future-releases/constant-update-2018/feature-specification.md]: Done.

File: accepted/future-releases/constant-update-2018/implementation-plan.md (hunk 6 above)
Comment (lrhn) [reply accepted/future-releases/constant-update-2018/implementation-plan.md]: Ack. But there is a phase 4, let's use that as 3 instead :)

**Pull request merged**